File: Semester2/B-CPE-200/corewar/corewar/srcs/parsing.c
/*
** EPITECH PROJECT, 2023
** B-CPE-200-STG-2-1-corewar-florent.guittre
** File description:
** parsing.c
*/

#include "vm.h"

/*
** In this .c file, we gonna parse the data + check for errors
** We to get:
**         - COREWAR_EXEC_MAGIC
**        - Name of the .cor
**        - Comments
**        - Instructions
*/

void ft_check_coding_byte(t_env *e, int k)
{
    char *str = e->player[k].string;
    int number = ((MM(str[0]) << 24) | (MM(str[1]) << 16) | (MM(str[2]) << 8) |
    MM(str[3]));

    if (number != COREWAR_EXEC_MAGIC)
        ft_exit(e, 4);
}

void ft_check_label_name(t_env *e, int k)
{
    int i = -1;
    char *name = e->player[k].name;

    while (name[++i])
        if (!(ft_strchr(CMD_CHARS, name[i])))
            ft_exit(e, 5);
}

void ft_check_label_comment(t_env *e, int k)
{
    int i = -1;
    char *comment = e->player[k].comment;

    while (comment[++i])
        if (!(ft_strchr(CMD_CHARS, comment[i])))
            ft_exit(e, 10);
}

void ft_check_all(t_env *e, int i)
{
    ft_check_coding_byte(e, i);
    ft_add_player_name(e, i);
    ft_add_player_comment(e, i);
    ft_check_label_name(e, i);
    ft_check_label_comment(e, i);
}

void ft_parsing(t_env *e)
{
    int i = -1;
    int k = 0;
    char *str = NULL;
    char *inst = NULL;

    while (++i < e->player_amount) {
        str = e->player[i].string;
        ft_check_all(e, i);
        k = PROG_NAME_LENGTH + COMMENT_LENGTH + 16;
        inst = e->player[i].inst;
        e->player[i].inst_len = k;
        while (k < e->player[i].len) {
            *inst = str[k];
            k++;
            inst++;
        }
        e->player[i].inst_len = k - e->player[i].inst_len;
        if (e->player[i].inst_len > CHAMP_MAX_SIZE)
            ft_exit(e, 6);
    }
}

File: Semester1/B-MAT-100/105torus/parsing.py
##
## EPITECH PROJECT, 2022
## B-MAT-100-STG-1-1-105torus-florent.guittre
## File description:
## parsing
##

import sys

def get_args():
    argv = sys.argv
    argc = len(argv)
    if (argc != 8):
        sys.exit(84)
    try:
        opt = int(argv[1])
        a0 = int(argv[2])
        a1 = int(argv[3])
        a2 = int(argv[4])
        a3 = int(argv[5])
        a4 = int(argv[6])
        n = int(argv[7])
    except:
        sys.exit(84)
    if (opt < 1 or opt > 3):
        sys.exit(84)
    if (n < 0):
        sys.exit(84)
    return (opt, a0, a1, a2, a3, a4, n)
